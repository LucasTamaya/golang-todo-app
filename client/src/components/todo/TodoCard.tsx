import { useState } from "react";

import ITodo from "../../interfaces/todo";
import checkIcon from "../../assets/img/check_icon.svg";
import uncheckIcon from "../../assets/img/uncheck_icon.svg";
import TodoDetails from "./TodoDetails";

const baseStyle =
  "flex flex-row justify-between items-center border rounded cursor-pointer p-2";

const Todo: React.FC<ITodo> = ({ title, body, done, id }) => {
  const [showDetails, setShowDetails] = useState(false);

  return (
    <>
      <li
        className={`${baseStyle} ${done ? "border-teal-400" : ""}`}
        onClick={() => setShowDetails((prev) => !prev)}
      >
        <p className={`${done ? "line-through" : ""}`}>{title}</p>
        <img
          src={done ? checkIcon : uncheckIcon}
          alt="check icon"
          width={20}
          height={20}
        />
      </li>

      {showDetails ? (
        <TodoDetails
          id={id}
          title={title}
          body={body}
          done={done}
          setShowDetails={setShowDetails}
        />
      ) : null}
    </>
  );
};

export default Todo;
