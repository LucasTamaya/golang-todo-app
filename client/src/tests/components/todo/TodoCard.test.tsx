import { fireEvent, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import { useState as useStateMock } from "react";

import TodoCard from "../../../components/todo/TodoCard";

// mock TodoDetails because of the api request
jest.mock("../../../components/todo/TodoDetails", () => {
  return null;
});

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

describe("TodoCard component", () => {
  const setState = jest.fn();

  beforeEach(() => {
    (useStateMock as jest.Mock).mockImplementation((init: false) => [
      init,
      setState,
    ]);
  });

  it("should renders the component correctly", () => {
    render(
      <TodoCard id="abc" title="Todo 1" body="Body of the todo" done={false} />
    );

    expect(screen.getByRole("listitem")).toBeInTheDocument();
    expect(screen.getByText(/todo 1/i)).toBeInTheDocument();
  });

  it("should match the snapshot", () => {
    const component = renderer.create(
      <TodoCard id="abc" title="Todo 1" body="Body of the todo" done={false} />
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should call setShowDetails if we click on the todo", () => {
    render(
      <TodoCard id="abc" title="Todo 1" body="Body of the todo" done={false} />
    );

    fireEvent.click(screen.getByRole("listitem"));

    expect(setState).toHaveBeenCalledTimes(1);
  });
});
