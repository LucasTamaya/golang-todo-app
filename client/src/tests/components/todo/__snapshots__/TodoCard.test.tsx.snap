// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoCard component should match the snapshot 1`] = `
<li
  className="flex flex-row justify-between items-center border rounded cursor-pointer p-2 "
  onClick={[Function]}
>
  <p
    className=""
  >
    Todo 1
  </p>
  <img
    alt="check icon"
    height={20}
    src="uncheck_icon.svg"
    width={20}
  />
</li>
`;
